src/Games/NexusMods.Games.MountAndBlade2Bannerlord/LauncherManager/NotificationProvider.cs

```diff
@@ -0,0 +1,51 @@
+using System.Collections.Concurrent;
+using Avalonia.Controls;
+using Bannerlord.LauncherManager.External.UI;
+using Bannerlord.LauncherManager.Localization;
+using Bannerlord.LauncherManager.Models;
+
+namespace NexusMods.Games.MountAndBlade2Bannerlord.LauncherManager;
+
+internal class NotificationProvider : INotificationProvider
+{
+    private readonly ConcurrentDictionary<string, object?> _notificationIds = new();
+    
+    private Window? _window; // TODO: How to inject the window?
+
+    public void SendNotification(string id, NotificationType type, string message, uint displayMs)
```

Comment:
We don't have a built-in notifications provider or abstractions for creating windows but if you need to show something that 'pops up' in front of the user, then the `Overlay` system.

(This is the system used for confirmation dialogs, such as login, deleting mods, etc.)

The caveat though is dismissing it requires user interaction.

To use that, you would use `IOverlayController` from `NexusMods.App.UI.Overlays`. There are some generic ones available like `MessageBoxOkView`. A good example of this is `ShowGameAlreadyRunningError` method. Would just need to lift `IOverlayController` into an abstractions package, I can do that if needed.

If there is a requirement to show info that doesn't require user interaction, we'd need to talk with UX (@captainsandypants). That's not currently in the App's designs.

Reply:
CC. @Nexus-Mods/nexusmods-app-developers 

We may want to have some sort of abstraction system for displaying UI elements from game modules, and external code. 

I think this was brought up around start of this year, back when we talked about emitting messages, as well as mediatR pattern and all.

Although we've managed to avoid it so far, there will be a time when game/external code will need to have some specific logic that need to render UI elements. I suppose that time is now.

----------

I guess some examples of messages emitted here (by @Aragas) would be helpful too, as this sort of thing may require design work too.

Reply:
Here's some examples
* `SendNotification("module-parsing-error-exception", NotificationType.Error, $"Failed to read SubModule.xml at path '{subModulePath}'!\n{e}", 3000);`
This one is triggered by the LauncherManager becuase it's reading the `/Modules` folder to get all mods. This is a basic operation, so if this fails we give a hint to the user

* `SendNotification("module-parsing-error-invalid-module-id", NotificationType.Error, $"SubModule.xml has an invalid Id at path '{subModulePath}'!", 3000);`
Same

* `SendNotification("sort-finished", NotificationType.Info, new BUTRTextObject("{=J7dh36Dy}Finished sorting!").ToString(), 3000);`
We expect the user to click on a "Auto Sort" button that triggers an asynchronous flow. This will give a confirmation that the operation finished

Reply:
They are present in the Vortex extension

Reply:
Hmm, I see.

We do have a missing piece of UX for reporting general errors that happen at runtime; right now it's all only ever emitted to log. I would like to see emitting messages of this kind to also be possible in the design we'll eventually come up with. I'll make sure to bring this up if others don't see this, sooner rather than later.

Since there's a whole process of asking UX/UI to do it, then discussing it/getting feedback, and then adding the whole abstraction layer.

Reply:
We also have a dialog that most likely won't be triggered
```
BUTRLoader detected 0Harmony.dll inside the game's root bin folder!
This could lead to issues, remove it?
```

BUTRLoader was the name for the BLSE backend, not used much anymore. It triggered a simple dialog box with Yes/No buttons

There are some other examples, but they are from an "optional" feature.  
BLSE has the ability to import/export load orders in all formats custom launchers have used (and of course introducing it's [own format](https://xkcd.com/927/)). The feature uses Yes/No dialogs and FileOpen/FileSave dialogs